### ISSUE TITLE
Support Exporting YOLOv8 to ONNX with IR Version 8

### ISSUE THREAD

**Issue by lasindu-codegen** [ultralytics/ultralytics#19498]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I'm trying to export a YOLOv8 model to ONNX with an intermediate representation (IR) version of 8. By default, the export produces a model with IR version 9, which is incompatible with my deployment environment. I attempted to force opset 8 during export, but this led to errors—specifically with the `upsample_nearest2d` operator.

**Steps to Reproduce:**
1. Run the following code snippet:
   ```python
   import torch
   from ultralytics import YOLO

   # Load the YOLOv8 model
   model = YOLO("yolov8m.pt")  # use the pretraiend weight 

   # Attempt to export to ONNX with opset 8 (expected to target IR version 8)
   onnx_path = "yolov8m.onnx"
   model.export(format="onnx", opset=8, dynamic=True)

### Additional

_No response_

**Comment by UltralyticsAssistant** [comment]
👋 Hello @lasindu-codegen, thank you for reaching out and sharing your question about exporting YOLOv8 to ONNX with IR version 8 🚀! We recommend checking out our [Docs](https://docs.ultralytics.com/) for guidance on YOLO usage, including export formats and troubleshooting tips.

If this is a 🐛 Bug Report, please ensure you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) that highlights the issue, including the exact error messages and configuration steps you've taken. This will help us pinpoint the problem more efficiently.

If your request is a ❓ Question about custom usage or export scenarios, please include additional details about your development environment (e.g., Python version, PyTorch version, ONNX runtime version) as this can have an impact on compatibility. In addition, sharing the full error traceback would be helpful for troubleshooting.

In the meantime, ensure you are using the latest version of the `ultralytics` package. You can upgrade to the latest release by running:

```bash
pip install -U ultralytics
```

You can run YOLOv8 in a variety of environments. If you haven't already, we recommend testing your export code in one of these verified setups:

- **Notebooks** with free GPU access:    
   <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  
- **Google Cloud** Deep Learning VM. Refer to our [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/).
- **Amazon AWS** Deep Learning AMI. See the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).
- **Docker Image**. You can test using the prebuilt YOLO Docker image. Instructions are available in the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### Status  

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  

Ensure the latest CI tests are passing, as these validate the correct functionality of the repository. If the badge is green, the repository is in good health for compatibility across tested environments.

💡 Just a friendly note: This is an automated response to assist you with initial troubleshooting and guidance. Rest assured, an Ultralytics engineer will review your issue and provide additional support soon! 😊

**Comment by Laughing-q** [comment]
@lasindu-codegen this is because onnx with `opset=8` does not include the `upsample_nearest2d` operator, recommend you to use higher opset version.

**Comment by Y-T-G** [comment]
IR version is not connected to the opset version. You need to downgrade the `onnx` package to get a lower IR version:

https://onnxruntime.ai/docs/reference/compatibility.html

For IR version 8, you need the onnx version 1.14

**Comment by lasindu-codegen** [comment]
Thanks @Y-T-G 

I did the following, but still I'm exporting onnx model with IR9

`model.export(format="onnx", dynamic=True)`

My python venv 3.8
pip install onnxruntime==1.14 ultralytics

**Comment by Y-T-G** [comment]
Not onnxruntime. You need to downgrade onnx. But onnxruntime too if you want to load and make inference on the same system.

`pip install onnx==1.14 onnxruntime==1.15`

**Comment by zjykzj** [comment]
> Not onnxruntime. You need to downgrade onnx. But onnxruntime too if you want to load and make inference on the same system.
> 
> `pip install onnx==1.14 onnxruntime==1.15`

@lasindu-codegen @Y-T-G @Laughing-q Did you solve this problem? make `onnx==1.14` could not make `IR_version=9` 

```
(py38) $ python3 tt.py
IR Version: 9
Opset Import: [version: 12
]
Producer Name: pytorch
Producer Version: 1.12.1
(py38) $
(py38) $
(py38) $ python3
Python 3.8.20 (default, Oct  3 2024, 15:24:27)
[GCC 11.2.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>>
>>> torch.__version__
'1.12.1'
>>>
>>> import onnx
>>> onnx.__version__
'1.14.0'
>>>
```

The complete dependency environment is as follows,

```
$ yolo checks
Ultralytics 8.3.100 🚀 Python-3.8.20 torch-1.12.1 CUDA:0 (NVIDIA GeForce RTX 3090, 24260MiB)
Setup complete ✅ (48 CPUs, 125.6 GB RAM, 827.0/937.8 GB disk)

OS                  Linux-5.4.0-150-generic-x86_64-with-glibc2.17
Environment         Linux
Python              3.8.20
Install             git
Path                /data/miniconda3/envs/py38/lib/python3.8/site-packages/ultralytics
RAM                 125.56 GB
Disk                827.0/937.8 GB
CPU                 Intel Xeon Silver 4214 2.20GHz
CPU count           48
GPU                 NVIDIA GeForce RTX 3090, 24260MiB
GPU count           2
CUDA                11.3

numpy               ✅ 1.24.3<=2.1.1,>=1.23.0
matplotlib          ✅ 3.7.5>=3.3.0
opencv-python       ❌ (not installed)>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.10.0>=1.4.1
torch               ✅ 1.12.1>=1.8.0
torch               ✅ 1.12.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.13.1>=0.9.0
tqdm                ✅ 4.65.2>=4.64.0
psutil              ✅ 6.1.1
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.0.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.13>=2.0.0
```

It is interesting that `torch 1.12.1` defaults to hard coded `IR-version=7`, Here is a classification model information that I have customized for conversion

```
(py38) $ ls
model.onnx  tt.py
(py38) $ python3 tt.py
IR Version: 7
Opset Import: [version: 12
]
Producer Name: pytorch
Producer Version: 1.12.1
```

So what happened in `yolo export` @UltralyticsAssistant @glenn-jocher 

```
[2025-06-11 10:48:08,979: WARNING/ForkPoolWorker-32] Starting export from /data/zj/renjie_train/rj_train/tmp/TinyPhoneDetect6313_train_1749610061/train/weights/best.pt to /data/zj/renjie_train/rj_train/tmp/TinyPhoneDetect6313_train_1749610061/train/weights/best.onnx
[2025-06-11 10:48:08,979: WARNING/ForkPoolWorker-32] Running command: yolo export model=/data/zj/renjie_train/rj_train/tmp/TinyPhoneDetect6313_train_1749610061/train/weights/best.pt format=onnx opset=12
[2025-06-11 10:48:14,294: WARNING/ForkPoolWorker-32] Command Output:
[2025-06-11 10:48:14,295: WARNING/ForkPoolWorker-32]
[2025-06-11 10:48:14,295: WARNING/ForkPoolWorker-32] Ultralytics 8.3.100 🚀 Python-3.8.20 torch-1.12.1 CPU (Intel Xeon Silver 4214 2.20GHz)
Model summary (fused): 72 layers, 3,006,233 parameters, 0 gradients, 8.1 GFLOPs

PyTorch: starting from '/data/zj/renjie_train/rj_train/tmp/TinyPhoneDetect6313_train_1749610061/train/weights/best.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 7, 8400) (5.9 MB)

ONNX: starting export with onnx 1.14.0 opset 12...
ONNX: slimming with onnxslim 0.1.34...
ONNX: export success ✅ 3.0s, saved as '/data/zj/renjie_train/rj_train/tmp/TinyPhoneDetect6313_train_1749610061/train/weights/best.onnx' (11.7 MB)

Export complete (3.3s)
Results saved to /data/zj/renjie_train/rj_train/tmp/TinyPhoneDetect6313_train_1749610061/train/weights
Predict:         yolo predict task=detect model=/data/zj/renjie_train/rj_train/tmp/TinyPhoneDetect6313_train_1749610061/train/weights/best.onnx imgsz=640
Validate:        yolo val task=detect model=/data/zj/renjie_train/rj_train/tmp/TinyPhoneDetect6313_train_1749610061/train/weights/best.onnx imgsz=640 data=/data/zj/renjie_train/rj_train/tmp/TinyPhoneDetect6313_train_1749610061/data_config.yaml
Visualize:       https://netron.app
💡 Learn more at https://docs.ultralytics.com/modes/export
```

**Comment by Y-T-G** [comment]
You can check the compatibility table to see which version corresponds to IR version 9.

https://onnxruntime.ai/docs/reference/compatibility.html

**Comment by zjykzj** [comment]
> You can check the compatibility table to see which version corresponds to IR version 9.
> 
> https://onnxruntime.ai/docs/reference/compatibility.html

Thank you @Y-T-G  for your reply, The link below can solve my problem #16959

* https://github.com/ultralytics/ultralytics/issues/16959#issuecomment-2416464815

```
import onnx

model_path = 'your_model.onnx'
model = onnx.load(model_path)
print(f"Current IR version: {model.ir_version}")

new_ir_version = 7
model.ir_version = new_ir_version

updated_model_path = 'updated_model.onnx'
onnx.save(model, updated_model_path)

print(f"IR version updated to: {new_ir_version}")
```
